File: src/main/java/com/joraph/schema/BasicCompositeKey.java
package com.joraph.schema;

import java.util.Arrays;
import java.util.function.Function;

public class BasicCompositeKey {

	public static final Function<Object[], BasicCompositeKey> CONVERTER = BasicCompositeKey::new;
	public static final Function<BasicCompositeKey, Object[]> CONVERTER_R = BasicCompositeKey::getParts;

	private Object[] objects;

	public BasicCompositeKey(Object... objects) {
		this.objects = objects;
	}

	public BasicCompositeKey(int size) {
		this.objects = new Object[size];
	}

	public Object getPart(int index) {
		return this.objects[index];
	}

	public void setPart(int index, Object object) {
		this.objects[index] = object;
	}

	public int getSize() {
		return this.objects.length;
	}

	public Object[] getParts() {
		return this.objects;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + Arrays.hashCode(objects);
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		BasicCompositeKey other = (BasicCompositeKey) obj;
		if (!Arrays.equals(objects, other.objects))
			return false;
		return true;
	}

}

File: src/main/kotlin/com/joraph/schema/BasicCompositeKey.kt
package com.joraph.schema


class BasicCompositeKey {

    companion object {

        @JvmField
        val CONVERTER: ((Array<Any?>) -> Any) = ::BasicCompositeKey
    }

    val parts: Array<Any?>
    val size: Int get() = parts.size

    constructor(vararg objects: Any?) {
        parts = arrayOf(*objects)
    }

    constructor(size: Int) {
        parts = arrayOfNulls(size)
    }

    operator fun get(index: Int): Any? = parts[index]

    operator fun set(index: Int, value: Any?) {
        parts[index] = value
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as BasicCompositeKey
        if (!parts.contentEquals(other.parts)) return false
        return true
    }

    override fun hashCode(): Int {
        return parts.contentHashCode()
    }

}
